contrib/babelfishpg_tds/test/t/004_bbfdumprestore.pl

```diff
@@ -0,0 +1,175 @@
+$tsql_oldnode->init_tsql('test_master', 'testdb');
+$oldnode->stop;
+
+# Initialize a new node for the restore.
+my $newnode = PostgreSQL::Test::Cluster->new('new_node');
+$newnode->init;
+$newnode->append_conf(
+	'postgresql.conf', qq{
+	log_connections = on
+	listen_addresses='127.0.0.1'
+	shared_preload_libraries = 'babelfishpg_tds'
+	lc_messages = 'C'
+});
+$newnode->start;
+# Initialize Babelfish in new node
+my $tsql_newnode = new TDSNode($newnode);
+$tsql_newnode->init_tsql('test_master', 'testdb');
+$newnode->stop;
+
+# Dump global objects using pg_dumpall. Note that we
+# need to use dump utilities from the new node here.
+$oldnode->start;
+my @dumpall_command = (
+	'pg_dumpall', '--database', 'testdb', '--username', 'test_master',
+	'--port', $oldnode->port, '--globals-only', '--quote-all-identifiers',
+	'--verbose', '--no-role-passwords', '--file', $dump1_file);
+$newnode->command_ok(\@dumpall_command, 'Dump global objects.');
+# Dump Babelfish database using pg_dump.
+my @dump_command = (
+	'pg_dump', '--username', 'test_master', '--quote-all-identifiers',
+	'--port', $oldnode->port, '--verbose', '--dbname', 'testdb',
+	'--file', $dump2_file);
+$newnode->command_ok(\@dump_command, 'Dump Babelfish database.');
+$oldnode->stop;
+
+# Retore the dumped files on the new server.
+$newnode->start;
+
+# Restore of dumpall file should cause a failure since cross version
+# dump/restore is not yet supported.
+$newnode->command_fails_like(
+	[
+		'psql',
+		'-d',         'testdb',
+		'-U',         'test_master',
+		'-p',         $newnode->port,
+		'-f',         $dump1_file,
+	],
+	qr/Backup and restore across different Postgres versions is not yet supported./,
+	'Restore of global objects failed since source and target versions do not match.');
+
+# Similarly, restore of dump file should also cause a failure.
+$newnode->command_fails_like(
+	[
+		'psql',
+		'-d',         'testdb',
+		'-U',         'test_master',
+		'-p',         $newnode->port,
+		'-f',         $dump2_file,
+	],
+	qr/Backup and restore across different Postgres versions is not yet supported./,
+	'Restore of Babelfish database failed since source and target versions do not match.');
+$newnode->stop;
+
+############################################################################################
+############################## Test for cross migration mode ###############################
+############################################################################################
+
+# Initialize a node with the current version to dump.
+my $newnode2 = PostgreSQL::Test::Cluster->new('new_node2');
+$newnode2->init;
+$newnode2->append_conf(
+	'postgresql.conf', qq{
+	log_connections = on
+	listen_addresses='127.0.0.1'
+	shared_preload_libraries = 'babelfishpg_tds'
+	lc_messages = 'C'
+});
+$newnode2->start;
+# Initialize Babelfish in new node
+my $tsql_newnode2 = new TDSNode($newnode2);
+$tsql_newnode2->init_tsql('test_master', 'testdb', 'multi-db');
```

Comment:
where are we making modes diff?

Reply:
So here we are passing `multi-db` for `newnode2` which we make it a multi-db server and `newnode1` (at line 53) is already started with `single-db` mode (default is `single-db` if we don't specify anything to `init_tsql`).